src/drivers/wa-sqlite/adapter.ts

```diff
@@ -0,0 +1,180 @@
+import {
+  DatabaseAdapter as DatabaseAdapterInterface,
+  RunResult,
+  Transaction as Tx,
+} from '../../electric/adapter'
+import { Database } from './database'
+import { parseTableNames, QualifiedTablename, Row, Statement } from '../../util'
+import { resultToRows } from '../absurd-sql'
+import { isInsertUpdateOrDeleteStatement } from '../../util/statements'
+import { Mutex } from 'async-mutex'
+
+export class DatabaseAdapter implements DatabaseAdapterInterface {
+  db: Database
+  txMutex: Mutex
+
+  constructor(db: Database) {
+    this.db = db
+    this.txMutex = new Mutex()
+  }
+
+  async runInTransaction(...statements: Statement[]): Promise<RunResult> {
+    // Uses a mutex to ensure that transactions are not interleaved.
+    // This is needed because transactions are executed manually using `BEGIN` and `COMMIT` commands.
+    const release = await this.txMutex.acquire()
```

Comment:
Yes, agreed. I was planning of looking into the other drivers at some point to introduce mutexes because even though some drivers support nested transactions, if two transactions get nested just because they happened to occur concurrently, that is not the semantics we want to achieve.

Reply:
okay. cool. That might even be related with the issue you were having on satellite. Please open a ticket and reference this discussion.